File: androidApp/src/main/java/com/badmintontracker/android/ui/theme/ShuttlType.kt
package com.badmintontracker.android.ui.theme

import androidx.compose.material3.Typography
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.Font
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.sp
import com.badmintontracker.android.R

/**
 * The type scale as raw numbers.
 *
 * Split out from the TextStyles so it can be asserted from a unit test with no
 * font resource and no Android runtime. Mirrors iosApp's ShuttlType.Role number
 * for number.
 *
 * Tracking is stored in em, the unit the design is expressed in, and converted
 * on the way out; storing sp would mean re-deriving it by hand on every size
 * change.
 */
internal object ShuttlScale {
    data class Role(
        val sizeSp: Float,
        val weight: FontWeight,
        val trackingEm: Float,
        val lineHeightMultiple: Float,
    ) {
        val trackingSp: Float get() = sizeSp * trackingEm
    }

    val display        = Role(40f, FontWeight.Medium, -0.035f, 1.08f)
    val headlineLarge  = Role(28f, FontWeight.Medium, -0.030f, 1.15f)
    val headlineMedium = Role(22f, FontWeight.Medium, -0.020f, 1.20f)
    val statNumber     = Role(26f, FontWeight.Medium, -0.030f, 1.15f)
    val wordmark       = Role(24f, FontWeight.Bold, -0.010f, 1.20f)
    val titleLarge     = Role(16f, FontWeight.SemiBold, -0.010f, 1.30f)
    val titleMedium    = Role(15f, FontWeight.SemiBold, -0.010f, 1.30f)
    val labelMedium    = Role(13f, FontWeight.SemiBold, -0.010f, 1.30f)
    val bodyLarge      = Role(16f, FontWeight.Normal, 0f, 1.45f)
    val bodyMedium     = Role(14f, FontWeight.Normal, 0f, 1.45f)
    val bodySmall      = Role(12f, FontWeight.Normal, 0f, 1.40f)
    val labelSmall     = Role(11f, FontWeight.Medium, 0.050f, 1.30f)
}

internal val Archivo = FontFamily(
    Font(R.font.archivo_regular, FontWeight.Normal),
    Font(R.font.archivo_medium, FontWeight.Medium),
    Font(R.font.archivo_semibold, FontWeight.SemiBold),
    Font(R.font.archivo_bold, FontWeight.Bold),
)

private fun ShuttlScale.Role.toTextStyle() = TextStyle(
    fontFamily    = Archivo,
    fontWeight    = weight,
    fontSize      = sizeSp.sp,
    lineHeight    = (sizeSp * lineHeightMultiple).sp,
    letterSpacing = trackingSp.sp,
)

private val Default = Typography()

/**
 * M3's default metrics with the family swapped to Archivo. Used for the six
 * [Typography] slots the design mock does not specify a size for
 * (displayLarge/Medium/Small, headlineSmall, titleSmall, labelLarge).
 * Inventing sizes for them would be design work this task has no mandate
 * for; keeping M3's own metrics and only changing the typeface still gets
 * the whole app off the system font, which is what this task is actually
 * for.
 */
private fun TextStyle.archivo() = copy(fontFamily = Archivo)

internal val ShuttlTypography = Typography(
    displayLarge   = Default.displayLarge.archivo(),
    displayMedium  = Default.displayMedium.archivo(),
    displaySmall   = Default.displaySmall.archivo(),
    headlineLarge  = ShuttlScale.headlineLarge.toTextStyle(),
    headlineMedium = ShuttlScale.headlineMedium.toTextStyle(),
    headlineSmall  = Default.headlineSmall.archivo(),
    titleLarge     = ShuttlScale.titleLarge.toTextStyle(),
    titleMedium    = ShuttlScale.titleMedium.toTextStyle(),
    titleSmall     = Default.titleSmall.archivo(),
    bodyLarge      = ShuttlScale.bodyLarge.toTextStyle(),
    bodyMedium     = ShuttlScale.bodyMedium.toTextStyle(),
    bodySmall      = ShuttlScale.bodySmall.toTextStyle(),
    labelLarge     = Default.labelLarge.archivo(),
    labelMedium    = ShuttlScale.labelMedium.toTextStyle(),
    // Tiny uppercase tracked label. The uppercasing is the caller's job, the
    // tracking is this style's.
    labelSmall     = ShuttlScale.labelSmall.toTextStyle(),
)

/**
 * Roles M3's [Typography] has no slot for. Phase 2's hero uses [display];
 * phase 3's stat tiles use [statNumber]; the sign-in brand mark uses
 * [wordmark]. [labelMedium] has an M3 slot of its own now (see
 * [ShuttlTypography]), so it is not duplicated here.
 */
internal object ShuttlTypeExtras {
    val display     = ShuttlScale.display.toTextStyle()
    val statNumber  = ShuttlScale.statNumber.toTextStyle()
    val wordmark    = ShuttlScale.wordmark.toTextStyle()
}

File: iosApp/Sources/Theme/ShuttlType.swift
import SwiftUI
import UIKit

/// The type scale, in Archivo.
///
/// Roles carry their raw numbers so the scale can be asserted from a unit test,
/// which a built `Font` cannot be. Mirrors androidApp's ShuttlScale number for
/// number; the two are checked against each other by ShuttlTypeTests and
/// ShuttlTypeTest.
///
/// Tracking is stored in em, the unit the design is expressed in, and converted
/// to points on the way out. Storing points would mean re-deriving the
/// conversion by hand every time a size changes.
enum ShuttlType {
    enum Weight: String, Hashable {
        case regular  = "Archivo-Regular"
        case medium   = "Archivo-Medium"
        case semibold = "Archivo-SemiBold"
        case bold     = "Archivo-Bold"
    }

    struct Role {
        let name: String
        let size: CGFloat
        let weight: Weight
        let trackingEm: CGFloat
        let lineHeightMultiple: CGFloat

        /// Tracking in points, which is what `.kerning` takes.
        var kerning: CGFloat { size * trackingEm }

        /// Extra leading in points, which is what `.lineSpacing` takes.
        ///
        /// Clamped at zero: `.lineSpacing` adds to the font's natural line
        /// height and cannot subtract from it, so a target tighter than Archivo's
        /// own leading is unreachable this way. See `View.shuttlType(_:)`.
        var lineSpacing: CGFloat {
            let natural = UIFont(name: weight.rawValue, size: size)?.lineHeight
                ?? UIFont.systemFont(ofSize: size).lineHeight
            return max(0, size * lineHeightMultiple - natural)
        }

        /// The same role at a different size, which is androidApp's
        /// `style.copy(fontSize = ...)`.
        ///
        /// For the handful of places where the scale's step is wrong but its
        /// face and its tracking are right - the number inside a 26pt progress
        /// ring, where labelSmall's 11pt does not leave room for three digits.
        /// Tracking follows the new size, because it is a fraction of it.
        func at(size: CGFloat) -> Role {
            Role(
                name: "\(name)@\(Int(size))",
                size: size,
                weight: weight,
                trackingEm: trackingEm,
                lineHeightMultiple: lineHeightMultiple
            )
        }

        /// The SwiftUI font. Falls back to the system font at the same size if
        /// Archivo is somehow missing, so a bundling mistake degrades rather
        /// than crashes. ArchivoFontTests is what catches it properly.
        var font: Font {
            .custom(weight.rawValue, size: size)
        }
    }

    static let display          = Role(name: "display", size: 40, weight: .medium, trackingEm: -0.035, lineHeightMultiple: 1.08)
    static let headlineLarge    = Role(name: "headlineLarge", size: 28, weight: .medium, trackingEm: -0.030, lineHeightMultiple: 1.15)
    static let headlineMedium   = Role(name: "headlineMedium", size: 22, weight: .medium, trackingEm: -0.020, lineHeightMultiple: 1.20)
    static let statNumber       = Role(name: "statNumber", size: 26, weight: .medium, trackingEm: -0.030, lineHeightMultiple: 1.15)
    static let wordmark         = Role(name: "wordmark", size: 24, weight: .bold, trackingEm: -0.010, lineHeightMultiple: 1.20)
    static let titleLarge       = Role(name: "titleLarge", size: 16, weight: .semibold, trackingEm: -0.010, lineHeightMultiple: 1.30)
    static let titleMedium      = Role(name: "titleMedium", size: 15, weight: .semibold, trackingEm: -0.010, lineHeightMultiple: 1.30)
    static let labelMedium      = Role(name: "labelMedium", size: 13, weight: .semibold, trackingEm: -0.010, lineHeightMultiple: 1.30)
    static let bodyLarge        = Role(name: "bodyLarge", size: 16, weight: .regular, trackingEm: 0, lineHeightMultiple: 1.45)
    static let bodyMedium       = Role(name: "bodyMedium", size: 14, weight: .regular, trackingEm: 0, lineHeightMultiple: 1.45)
    static let bodySmall        = Role(name: "bodySmall", size: 12, weight: .regular, trackingEm: 0, lineHeightMultiple: 1.40)
    static let labelSmall       = Role(name: "labelSmall", size: 11, weight: .medium, trackingEm: 0.050, lineHeightMultiple: 1.30)

    static let allRoles: [Role] = [
        display, headlineLarge, headlineMedium, statNumber, wordmark,
        titleLarge, titleMedium, labelMedium, bodyLarge, bodyMedium, bodySmall, labelSmall,
    ]
}

extension Text {
    /// A role applied to text that will be drawn inside a `Canvas`.
    ///
    /// `GraphicsContext.resolve` takes a `Text`, not a view, so the `View`
    /// modifier below cannot reach it - and a canvas label left to set its own
    /// `.font` is how a drawing quietly reverts to San Francisco while every
    /// screen around it is on Archivo.
    ///
    /// Leading is the one part of a role a canvas label cannot carry: a resolved
    /// `Text` is positioned by the drawing, not laid out in a stack, so there is
    /// no line spacing to set. Every canvas label in this app is a single line.
    ///
    /// Named apart from `shuttlType(_:)` deliberately. An overload on `Text`
    /// returning `Text` would win resolution at every ordinary `Text(...)
    /// .shuttlType(...)` call site in the app and silently drop the leading
    /// those DO need.
    func shuttlCanvasType(_ role: ShuttlType.Role) -> Text {
        self.font(role.font).kerning(role.kerning)
    }
}

extension View {
    /// Applies a role's font, tracking and leading together, which is the only
    /// correct way to use one: kerning and leading are both derived from the
    /// size, so setting the font alone silently ships the wrong metrics.
    ///
    /// Leading is approximate in one direction. SwiftUI's `.lineSpacing` is
    /// ADDITIVE - it adds to the font's natural line height and cannot subtract
    /// from it - so a role whose target sits below Archivo's natural leading
    /// gets that natural leading instead, and reads slightly looser than the
    /// design. Only the display role is affected. The hero is the one place
    /// where that difference is visible, and phase 2 draws it as two separate
    /// `Text` views in a `VStack` with explicit spacing, which sidesteps
    /// leading entirely and hits 1.08 exactly.
    func shuttlType(_ role: ShuttlType.Role) -> some View {
        self.font(role.font)
            .kerning(role.kerning)
            .lineSpacing(role.lineSpacing)
    }

    /// Same as `shuttlType(_:)`, with monospaced digits - for timers and counts
    /// that must not shift width as their digits change.
    func shuttlType(_ role: ShuttlType.Role, monospacedDigit: Bool) -> some View {
        self.font(monospacedDigit ? role.font.monospacedDigit() : role.font)
            .kerning(role.kerning)
            .lineSpacing(role.lineSpacing)
    }
}
